File: FlowerObjc/Source/UIKit/SGTextView.m
//
//  SGTextView.m
//  SGTextView
//
//  Created by kingsic on 2020/12/11.
//  Copyright © 2020 kingsic. All rights reserved.
//

#import "SGTextView.h"

@interface SGTextView ()
@property (nonatomic, strong) UILabel *placeholderLabel;
@property (nonatomic, assign) BOOL addPlaceholderNotification;
@property (nonatomic, assign) BOOL addLimitNumberNotification;
@end

@implementation SGTextView

- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat x = 6;
    CGFloat y = 0;
    CGFloat w = self.frame.size.width - 2 * x;
    CGFloat h = 0;
    if (self.addPlaceholderNotification) {
        if (self.font == nil) {
            y = 6.5;
            h = [self P_calculateHeightWithString:_placeholder font:_placeholderLabel.font width:w];
        } else {
            y = 8;
            h = [self P_calculateHeightWithString:_placeholder font:self.font width:w];
        }
        _placeholderLabel.frame = CGRectMake(x, y, w, h);
    }
}

- (UILabel *)placeholderLabel {
    if (!_placeholderLabel) {
        _placeholderLabel = [[UILabel alloc] init];
        _placeholderLabel.text = _placeholder;
        _placeholderLabel.numberOfLines = 0;
        _placeholderLabel.textColor = [UIColor lightGrayColor];
        if (self.font == nil) {
            _placeholderLabel.font = [UIFont systemFontOfSize:13];
        } else {
            _placeholderLabel.font = self.font;
        }
    }
    return _placeholderLabel;
}

- (void)setPlaceholder:(NSString *)placeholder {
    _placeholder = placeholder;
    
    [self addSubview:self.placeholderLabel];
    if (self.addPlaceholderNotification == NO && self.addLimitNumberNotification == NO) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textDidChangeNotification) name:UITextViewTextDidChangeNotification object:self];
        self.addPlaceholderNotification = YES;
    } else if (self.addPlaceholderNotification == NO && self.addLimitNumberNotification) {
        self.addPlaceholderNotification = YES;
    }
}
- (void)setPlaceholderColor:(UIColor *)placeholderColor {
    _placeholderColor = placeholderColor;
    _placeholderLabel.textColor = placeholderColor;
}

- (void)setLimitNumber:(NSInteger)limitNumber {
    _limitNumber = limitNumber;
    if (limitNumber <= 0) {
        @throw [NSException exceptionWithName:@"SGTextView" reason:@"【limitNumber属性值必须大于等于0】" userInfo:nil];;
    }
    if (self.addLimitNumberNotification == NO && self.addPlaceholderNotification == NO) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textDidChangeNotification) name:UITextViewTextDidChangeNotification object:self];
        self.addLimitNumberNotification = YES;
    } else if (self.addLimitNumberNotification == NO && self.addPlaceholderNotification) {
        self.addLimitNumberNotification = YES;
    }
}

- (void)textDidChangeNotification {
    if (self.addPlaceholderNotification) {
        _placeholderLabel.hidden = self.hasText;
    }
    if (self.addLimitNumberNotification) {
        if (_limitNumber > 0) {
            if (self.text.length > _limitNumber) {
                self.text = [self.text substringToIndex:_limitNumber];
            }
        }
    }
}

- (void)dealloc {
    if (self.addPlaceholderNotification || self.addLimitNumberNotification) {
        [[NSNotificationCenter defaultCenter] removeObserver:self];
    }
}


/**
 *  根据字号大小、宽度及字符串上下间的间距计算出字符串的高
 */
- (CGFloat)P_calculateHeightWithString:(NSString *)string font:(UIFont *)font width:(CGFloat)width {
    NSDictionary *attributes = @{NSFontAttributeName : font};
    return [string boundingRectWithSize:CGSizeMake(width, 0) options:NSStringDrawingUsesLineFragmentOrigin attributes:attributes context:nil].size.height;
}

@end

File: Flower/Sources/SGTextView/SGTextView.swift
//
//  SGTextView.swift
//  Flower
//
//  Created by kingsic on 2020/9/22.
//  Copyright © 2020 kingsic. All rights reserved.
//

import UIKit

public class SGTextView: UITextView  {
    /// PlaceHolder
    public var placeHolder: String? {
        didSet {
            placeHolderLabel.text = placeHolder
        }
    }
    
    /// PlaceHolder Color
    public var placeHolderColor: UIColor? {
        didSet {
            placeHolderLabel.textColor = placeHolderColor
        }
    }
    
    /// Limit input words
    public var limitNumber: Int?
    
    /// PlaceHolder Label 
    private lazy var placeHolderLabel: UILabel = {
        $0.font = UIFont.systemFont(ofSize: 12)
        $0.text = "请输入内容~"
        $0.textColor = UIColor.lightGray
        $0.numberOfLines = 0
        return $0
    }(UILabel())
    
    public override var font: UIFont? {
        didSet {
            if font != nil {
                // 让在属性哪里修改的字体,赋给给我们占位label
                placeHolderLabel.font = font
            }
        }
    }
    
    public override var text: String? {
        didSet {
            // 根据文本是否有内容而显示占位label
            placeHolderLabel.isHidden = hasText
        }
    }
    
    // frame
    override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        setupUI()
    }
    // xib
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setupUI()
    }
    
    // 添加控件, 设置约束
    fileprivate func setupUI() {
        // 监听内容的通知
        NotificationCenter.default.addObserver(self, selector: #selector(SGTextView.valueChange), name: UITextView.textDidChangeNotification, object: nil)
        // UITextViewTextDidChangeNotification
        // 添加控件
        addSubview(placeHolderLabel)
        
        // 设置约束,使用系统的约束
        placeHolderLabel.translatesAutoresizingMaskIntoConstraints = false
        
        addConstraint(NSLayoutConstraint(item: placeHolderLabel, attribute: .width, relatedBy: .equal, toItem: self, attribute: .width, multiplier: 1, constant: -10))
    }
    
    // 内容改变的通知方法
    @objc fileprivate func valueChange() {
        // 占位文字的显示与隐藏
        placeHolderLabel.isHidden = hasText
        
        // 字数限制处理
        if limitNumber != nil {

            guard limitNumber! > 0 else {
                return
            }
            
            if self.text!.count > limitNumber! {
                self.text = (self.text! as NSString).substring(to: limitNumber!)
            }
        }
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        // 设置占位文字的坐标
        placeHolderLabel.frame.origin.x = 5
        placeHolderLabel.frame.origin.y = 7
    }
    
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
}

